lib/rules/space-before-function-paren.js

```diff
@@ -92,13 +102,28 @@ module.exports = {
          * @returns {void}
          */
         function validateSpacingBeforeParentheses(node) {
-            const isNamed = isNamedFunction(node);
-            let rightToken;
+            const isArrow = node.type === "ArrowFunctionExpression";
+            const isNamed = !isArrow && isNamedFunction(node);
+            let forbidSpacing, requireSpacing, rightToken;
 
-            if (node.generator && !isNamed) {
+            if ((node.generator && !isNamed) ||
+                (isArrow && !node.async) ||
+                (isArrow && sourceCode.getFirstToken(node, 1).value !== "(")
```

Comment:
Checking my understanding, these checks are:

1) Un-named generator function expressions. This is covered by `generator-star-spacing`.
2) Synchronous arrow functions, which do not have a name.
3) Asynchronous arrow functions that have only a single parameter and omit the optional parens.

Comments summarizing these would be helpful, or maybe extract them to named helper functions?